### PR TITLE
Add pagination for user state API

### PR DESCRIPTION
**Pull request by Achintya-Chatterjee**
## Issue Ticket Number:- 
- Closes: https://github.com/Real-Dev-Squad/website-dashboard/issues/529
- Desgn Doc: https://docs.google.com/document/d/1KtwQPNylvsPMkQTtEc6FGLzR4xqR2dlepnpZr1XCVW4/edit
## Description: 
- Modify the website-backend code to support pagination and filtering based on the selected checkboxes.
- Handle the selected checkbox values in the website-backend code to filter the data accordingly. 
- I ensured that the API endpoint can accept ‘page’ and ‘size’ parameters.

Is Under the Feature Flag 
- [ ] Yes
- [x] No

Database changes
- [ ] Yes
- [x] No

Breaking changes (If your feature is breaking/missing something please mention pending tickets)
- [ ] Yes
- [x] No

Is Development Tested?

- [x] Yes
- [ ] No

### Add relevant Screenshot below ( e.g test coverage etc. )
Screenshot from 2024-01-31 00-07-56

## Without Feature Flag
Screenshot from 2024-02-02 11-43-46

## With Feature Flag
Screenshot from 2024-02-02 11-47-08

## Before:
Screenshot from 2024-01-28 16-15-36

## After:
Screenshot from 2024-01-31 00-52-32

Screenshot from 2024-02-02 14-22-23

## API contract:
https://github.com/Real-Dev-Squad/website-api-contracts/pull/176

### CODE REVIEW

File: controllers/users.js [hunk 1]
```diff
@@ -738,11 +738,34 @@ const filterUsers = async (req, res) => {
...
+    const totalCount = await users.length;
+    const totalPages = Math.ceil(totalCount / limitNumber);
+    const nextPage = pageNumber < totalPages - 1 ? pageNumber + 1 : null;
+    const prevPage = pageNumber > 0 ? pageNumber - 1 : null;
+
+    const paginationLinks = {
+      next: nextPage ? getFilteredPaginationLink(req.query, nextPage, limitNumber) : null,
+      prev: prevPage ? getFilteredPaginationLink(req.query, prevPage, limitNumber) : null,
+    };
+
```
Comment (yesyash) [inline controllers/users.js:756]: - can we create a function for this which returns us the required value for the pagination object below, this way the controller does not grow big?

File: controllers/users.js [hunk 2]
```diff
@@ -738,11 +738,34 @@ const filterUsers = async (req, res) => {
...
+    const users = await dataAccess.retreiveFilteredUsers(req.query, skip, limitNumber);
+    const totalCount = await users.length;
+    const totalPages = Math.ceil(totalCount / limitNumber);
+    const nextPage = pageNumber < totalPages - 1 ? pageNumber + 1 : null;
+    const prevPage = pageNumber > 0 ? pageNumber - 1 : null;
+
+    const paginationLinks = {
+      next: nextPage ? getFilteredPaginationLink(req.query, nextPage, limitNumber) : null,
+      prev: prevPage ? getFilteredPaginationLink(req.query, prevPage, limitNumber) : null,
```
Comment (yesyash) [inline controllers/users.js:754]: - what are we going to use this for ?

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: controllers/users.js [hunk 3]
```diff
@@ -738,11 +738,34 @@ const filterUsers = async (req, res) => {
       return res.boom.badRequest("filter for item not provided");
     }
 
-    const users = await dataAccess.retreiveFilteredUsers(req.query);
+    const { page, limit } = req.query;
+    const pageNumber = parseInt(page) || 0;
+    const limitNumber = parseInt(limit) || 100;
```
[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply controllers/users.js:743]: At the time giving the task, Ankush told me he wanted to see 100 data per page. If you say so I will change it back to 10.

File: controllers/users.js (hunk 2 above)
Comment (Achintya-Chatterjee) [reply controllers/users.js:754]: The **paginationLinks** object in my code prepares these links(next, prev). The **getFilteredPaginationLink** function is responsible for constructing the URL for the next or previous page based on the current query parameters.

This information is useful for clients (e.g., front-end applications) that want to fetch additional pages of data without having to retrieve the entire dataset in a single request.

File: controllers/users.js [hunk 4]
```diff
@@ -738,11 +738,34 @@ const filterUsers = async (req, res) => {
...
       users: users,
       count: users.length,
+      pagination: {
+        totalCount,
+        totalPages,
+        currentPage: pageNumber,
+        nextPage,
+        prevPage,
+        links: paginationLinks,
+      },
```
Comment (prakashchoudhary07) [inline controllers/users.js:768]: let's keep our pagination format the same throughout, please have a look at others API's where pagination is done

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: controllers/users.js (hunk 4 above)
Comment (Achintya-Chatterjee) [reply controllers/users.js:768]: cool, will look into it and I will try my level best to do it same way.

File: controllers/users.js (hunk 1 above)
Comment (Achintya-Chatterjee) [reply controllers/users.js]: created the function which will be returning the pagination object

File: controllers/users.js (hunk 4 above)
Comment (Achintya-Chatterjee) [reply controllers/users.js]: @prakashchoudhary07 changed it. Can you please look into it?

File: controllers/users.js [hunk 5]
```diff
@@ -732,16 +732,36 @@ const addDefaultArchivedRole = async (req, res) => {
...
+
 const filterUsers = async (req, res) => {
   try {
     if (!Object.keys(req.query).length) {
       return res.boom.badRequest("filter for item not provided");
     }
+    const { page, limit } = req.query;
+    const pageNumber = parseInt(page) || 0;
+    const limitNumber = parseInt(limit) || 100;
+    const skip = (pageNumber - 1) * limitNumber;
```
Comment (prakashchoudhary07) [inline controllers/users.js:753]: Do you know if the skip functionality is implemented?

Comment (Achintya-Chatterjee) [reply controllers/users.js:753]: Yes, It is implemented

`const users = await dataAccess.retreiveFilteredUsers(req.query, skip, limitNumber);`

**Review by iamitprakash — CHANGES_REQUESTED** [review]
please add test case stats

**Comment by Achintya-Chatterjee** [comment]
> please add test case stats

Added

**Comment by prakashchoudhary07** [comment]
@Achintya-Chatterjee Please share a result/screenshot of actual pagination result, happening
In the above result, there is only one record present, please add some more records and share the response with the paginated values

File: middlewares/validators/userStatus.js [hunk 6]
```diff
@@ -125,7 +125,6 @@ const validateGetQueryParams = async (req, res, next) => {
     .messages({
       "object.unknown": "Invalid query param provided.",
     });
-
```
Comment (ajoykumardas12) [inline middlewares/validators/userStatus.js:128]: This empty line was fine I think.

File: test/integration/usersFilter.test.js [hunk 7]
```diff
@@ -154,6 +154,28 @@ describe("Filter Users", function () {
...
+            return done(err);
+          }
+          expect(res).to.have.status(200);
+          expect(res.body).to.be.a("object");
+          expect(res.body).to.have.property("message");
+          expect(res.body).to.have.property("users");
+          expect(res.body).to.have.property("count");
+          expect(res.body).to.have.property("links"); // Corrected property name
+          expect(res.body.links).to.have.property("next"); // If applicable
+          expect(res.body.links).to.have.property("prev"); // If applicable
```
Comment (ajoykumardas12) [inline test/integration/usersFilter.test.js:174]: Didn't get the comments.

File: controllers/users.js (hunk 5 above)
Comment (ajoykumardas12) [reply controllers/users.js:753]: Can we have some test or proof of skip functionality? Didn't find this functionality in code.

**Comment by Achintya-Chatterjee** [comment]
> @Achintya-Chatterjee Please share a result/screenshot of actual pagination result, happening In the above result, there is only one record present, please add some more records and share the response with the paginated values



> @Achintya-Chatterjee Please share a result/screenshot of actual pagination result, happening In the above result, there is only one record present, please add some more records and share the response with the paginated values

cool, will update this tonight @prakashchoudhary07

**Review by yesyash — APPROVED** [review]

File: test/integration/usersFilter.test.js (hunk 7 above)
Comment (Achintya-Chatterjee) [reply test/integration/usersFilter.test.js]: removed the comments

File: middlewares/validators/userStatus.js (hunk 6 above)
Comment (Achintya-Chatterjee) [reply middlewares/validators/userStatus.js:128]: added

**Review by yesyash — DISMISSED** [review]

**Review by ajoykumardas12 — DISMISSED** [review]

File: controllers/users.js (hunk 5 above)
Comment (Achintya-Chatterjee) [reply controllers/users.js:753]: @prakashchoudhary07 @ajoykumardas12 added test for skip functionality.

File: test/integration/usersFilter.test.js [hunk 8]
```diff
@@ -177,6 +199,29 @@ describe("Filter Users", function () {
         });
     });
 
+    it("Should skip correct number of users", function (done) {
+      chai
+        .request(app)
+        .get("/users/search")
+        .query({ page: 1, size: 100 })
```
Comment (ajoykumardas12) [inline test/integration/usersFilter.test.js:206]: This test query doesn't include skip

**Review by ajoykumardas12 — APPROVED** [review]

**Review by yesyash — APPROVED** [review]

**Comment by prakashchoudhary07** [comment]
Are these changes behind a feature flag?

**Comment by Achintya-Chatterjee** [comment]
> Are these changes behind a feature flag?

@prakashchoudhary07 Nope

**Comment by prakashchoudhary07** [comment]
> > Are these changes behind a feature flag?
> 
> @prakashchoudhary07 Nope

Please add these changes behind a feature flag, 
We will always have new changes behind the feature flag

**Comment by Achintya-Chatterjee** [comment]
> > > Are these changes behind a feature flag?
> > 
> > 
> > @prakashchoudhary07 Nope
> 
> Please add these changes behind a feature flag, We will always have new changes behind the feature flag

@prakashchoudhary07 will add these tomorrow morning after the standup

**Comment by Achintya-Chatterjee** [comment]
> > > > Are these changes behind a feature flag?
> > > 
> > > 
> > > @prakashchoudhary07 Nope
> > 
> > 
> > Please add these changes behind a feature flag, We will always have new changes behind the feature flag
> 
> @prakashchoudhary07 will add these tomorrow morning after the standup

@prakashchoudhary07 done with the implementation. pagination is now behind feature flag

File: middlewares/validators/user.js [hunk 9]
```diff
@@ -207,38 +207,39 @@ async function getUsers(req, res, next) {
...
+      size: joi.number().integer().min(1).max(100).message("size must be a number between 1 and 100").optional(),
+      prev: joi.string().allow("").optional(),
+      next: joi.string().allow("").optional(),
+      dev: joi.bool().optional().sensitive(),
     })
-    .messages({
-      "object.min": "Please provide at least one filter criteria",
-    });
+    .min(1)
+    .messages({ "object.min": "Please provide at least one filter criterion" });
```
[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply middlewares/validators/user.js]: Done, Please check once you're free

File: controllers/users.js [hunk 10]
```diff
@@ -732,16 +732,45 @@ const addDefaultArchivedRole = async (req, res) => {
...
+        count: users.length,
+      });
+    }
+    const { page, size } = req.query;
+    const pageNumber = parseInt(page) || 0;
+    const limitNumber = parseInt(size) || 100;
+    const skip = (pageNumber - 1) * limitNumber;
+
+    const users = await dataAccess.retreiveFilteredUsers(req.query, skip, limitNumber);
+    const totalCount = await users.length;
```
[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply controllers/users.js:765]: The `await` keyword is used when calling `dataAccess.retreiveFilteredUsers` to wait for the asynchronous operation to complete before moving on to the next line of code.

File: controllers/users.js [hunk 11]
```diff
@@ -732,16 +732,45 @@ const addDefaultArchivedRole = async (req, res) => {
...
 const filterUsers = async (req, res) => {
   try {
     if (!Object.keys(req.query).length) {
       return res.boom.badRequest("filter for item not provided");
     }
+    const dev = req.query.dev;
+    if (dev !== "true") {
+      const users = await dataAccess.retreiveFilteredUsers(req.query);
+      return res.json({
+        message: users.length ? "Users found successfully!" : "No users found",
```
[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply controllers/users.js:754]: I had to refactor a lot of code for this change. can we do it a bit later?

File: controllers/users.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply controllers/users.js]: remove the `totalCount` await

[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply controllers/users.js]: The `await` keyword is not needed when we're calculating the `totalCount` we're just getting the length of the `users` array. This change will improve code readability and remove unnecessary use of `await`.

Thanks for pointing this out. @yesyash

File: controllers/users.js [hunk 12]
```diff
@@ -732,16 +732,45 @@ const addDefaultArchivedRole = async (req, res) => {
...
+const calculatePagination = (pageNumber, totalPages, reqQuery, limitNumber) => {
+  const nextPage = pageNumber < totalPages - 1 ? pageNumber + 1 : null;
+  const prevPage = pageNumber > 0 ? pageNumber - 1 : null;
+
+  return {
+    next: nextPage ? getFilteredPaginationLink(reqQuery, nextPage, limitNumber) : null,
+    prev: prevPage ? getFilteredPaginationLink(reqQuery, prevPage, limitNumber) : null,
+  };
+};
+
```
[the remark this answers is not in the document]
Comment (Achintya-Chatterjee) [reply controllers/users.js:744]: let it be for some time as it is. I will move this to util functions later. I was trying to move it there but got lots of errors.

File: controllers/users.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (yesyash) [reply controllers/users.js:754]: Sure

**Review by vinit717 — APPROVED** [review]

**Review by ajoykumardas12 — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]

**Pull request merged**